src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -0,0 +1,350 @@
+        val matchedSets = mutable.Set[UmiSet]()
+        for (set <- umiSets) {
+          if (set.exists(other => countMismatches(other, umi) <= maxMismatches)) {
+            set.add(umi)
+            matchedSets.add(set)
+          }
+        }
+
+        matchedSets.size match {
+          case 0 => umiSets += mutable.SortedSet(umi)
+          case 1 => matchedSets.head.add(umi)
+          case _ =>
+            // Merge the sets and remove all but one from umiSets
+            val pick = matchedSets.head
+            matchedSets.tail.foreach(set => {
+              umiSets -= set
+              pick ++= set
+            })
+        }
+      }
+
+      umiSets.flatMap(set => set.map(umi => umi -> set.head)).toMap
+    }
+  }
+
+  /** Class that implements the directed adjacency graph method from umi_tools. */
+  class AdjacencyUmiAssigner(val maxMismatches: Int) extends UmiAssigner {
+    /** Represents a node in the adjacency graph; equality is just by UMI sequence. */
+    private class Node(val umi: String, val count: Long, val children: mutable.Buffer[Node] = mutable.Buffer()) {
+      /** Gets the full set of descendants from this node. */
+      def descendants: List[Node] = {
+        val buffer = ListBuffer[Node]()
+        addDescendants(buffer)
+        buffer.toList
+      }
+
+      private def addDescendants(buffer: mutable.Buffer[Node]): Unit = children.foreach(child => {
+        buffer += child
+        child.addDescendants(buffer)
+      })
+
+      override def equals(other: scala.Any): Boolean = other.isInstanceOf[Node] && this.umi == other.asInstanceOf[Node].umi
+    }
+
+    override def assign(rawUmis: Seq[String]): Map[String, String] = {
+      // Make a list of counts of all UMIs in order from most to least abundant
+      val nodes = SimpleCounter(rawUmis).map(x => new Node(x._1, x._2)).toBuffer.sortBy((n:Node) => -n.count)
+      val roots = mutable.Buffer[Node]()
+
+      // Now build one or more graphs starting with the most abundant remaining umi
+      def addChildren(root: Node, others: mutable.Buffer[Node]) : Unit = {
+        val xs = others.filter(other => countMismatches(root.umi, other.umi) <= maxMismatches && root.count >= 2 * other.count - 1)
+        root.children ++= xs
+        others --= xs
+        root.children.foreach(r => addChildren(r, others))
+      }
+
+      while (nodes.nonEmpty) {
+        val root = nodes.remove(0)
+        addChildren(root, nodes)
+        roots += root
+      }
+
+      // And finally make the output mapping
+      val mappings = mutable.Buffer[(String,String)]()
+      roots.foreach(root => {
+        mappings.append((root.umi, root.umi))
+        root.children.foreach(child => mappings.append((child.umi, root.umi)))
+      })
+
+      mappings.toMap
+    }
+  }
+
+  /** Looks in all the places the library name can be hiding. Returns the library name
+    * if one is found, otherwise returns "unknown".
+    */
+  final def library(rec: SAMRecord): String = {
```

Comment:
Agreed.  Ideally this code would be inside SAMRecord.  I blame the spec for having multiple places to store `LB`.